sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -0,0 +1,151 @@
+
+import org.apache.hadoop.fs.{FileStatus, FileSystem, Path}
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.sql.types.{StringType, StructType}
+import org.apache.spark.util.collection.OpenHashSet
+
+/**
+ * A very simple source that reads text files from the given directory as they appear.
+ */
+class FileStreamSource(
+    sqlContext: SQLContext,
+    metadataPath: String,
+    path: String,
+    dataSchema: Option[StructType],
+    dataFrameBuilder: Array[String] => DataFrame) extends Source with Logging {
+
+  import sqlContext.implicits._
+
+  /** Returns the schema of the data from this source */
+  override lazy val schema: StructType = {
+    dataSchema.getOrElse {
+      val filesPresent = fetchAllFiles()
+      if (filesPresent.isEmpty) {
+        new StructType().add("value", StringType)
+      } else {
+        // There are some existing files. Use them to infer the schema
+        dataFrameBuilder(filesPresent.toArray).schema
+      }
+    }
+  }
+
+  /** Returns the maximum offset that can be retrieved from the source. */
+  def fetchMaxOffset(): LongOffset = synchronized {
+    val filesPresent = fetchAllFiles()
+    val newFiles = new ArrayBuffer[String]()
+    filesPresent.foreach { file =>
+      if (!seenFiles.contains(file)) {
+        logDebug(s"new file: $file")
+        newFiles.append(file)
+        seenFiles.add(file)
+      } else {
+        logDebug(s"old file: $file")
+      }
+    }
+
+    if (newFiles.nonEmpty) {
+      maxBatchFile += 1
+      writeBatch(maxBatchFile, newFiles)
+    }
+
+    new LongOffset(maxBatchFile)
+  }
+
+  def currentOffset: LongOffset = synchronized {
+    new LongOffset(maxBatchFile)
+  }
+
+  /**
+   * Returns the next batch of data that is available after `start`, if any is available.
+   */
+  override def getNextBatch(start: Option[Offset]): Option[Batch] = {
+    val startId = start.map(_.asInstanceOf[LongOffset].offset).getOrElse(-1L)
+    val end = fetchMaxOffset()
+    val endId = end.offset
+
+    val batchFiles = (startId + 1 to endId).filter(_ >= 0).map(i => s"$metadataPath/$i")
+    if (batchFiles.nonEmpty) {
+      logDebug(s"Producing files from batches ${startId + 1}:$endId")
+      logDebug(s"Batch files: $batchFiles")
+
+      // Probably does not need to be a spark job...
+      val files = sqlContext
+          .read
+          .text(batchFiles: _*)
+          .as[String]
+          .collect()
+      logDebug(s"Streaming ${files.mkString(", ")}")
+      Some(new Batch(end, dataFrameBuilder(files)))
+    } else {
+      None
+    }
+  }
+
+  private def sparkContext = sqlContext.sparkContext
+
+  private val fs = FileSystem.get(sparkContext.hadoopConfiguration)
+  private val existingBatchFiles = fetchAllBatchFiles()
+  private val existingBatchIds = existingBatchFiles.map(_.getPath.getName.toInt)
+  private var maxBatchFile = if (existingBatchIds.isEmpty) -1 else existingBatchIds.max
```

Comment:
nit: Isnt this `maxBatchId` rather than the file (since its an int, rather than a string)?